pkg/services/ngalert/state/manager.go

```diff
@@ -166,51 +167,63 @@ func (st *Manager) ProcessEvalResults(ctx context.Context, alertRule *ngModels.A
 }
 
 // Set the current state based on evaluation results
-func (st *Manager) setNextState(ctx context.Context, alertRule *ngModels.AlertRule, result eval.Result) *State {
-	currentState := st.getOrCreate(ctx, alertRule, result)
+func (st *Manager) setNextState(ctx context.Context, alertRule *ngModels.AlertRule, result eval.Result) *AlertInstance {
+	instance := st.getOrCreate(ctx, alertRule, result)
 
-	currentState.LastEvaluationTime = result.EvaluatedAt
-	currentState.EvaluationDuration = result.EvaluationDuration
-	currentState.Results = append(currentState.Results, Evaluation{
+	instance.LastEvaluationTime = result.EvaluatedAt
+	instance.EvaluationDuration = result.EvaluationDuration
+	instance.Results = append(instance.Results, Evaluation{
 		EvaluationTime:  result.EvaluatedAt,
 		EvaluationState: result.State,
 		Values:          NewEvaluationValues(result.Values),
 		Condition:       alertRule.Condition,
 	})
-	currentState.LastEvaluationString = result.EvaluationString
-	currentState.TrimResults(alertRule)
-	oldState := currentState.State
+	instance.LastEvaluationString = result.EvaluationString
+	instance.TrimResults(alertRule)
+	oldState := instance.State
+	oldReason := instance.StateReason
 
 	st.log.Debug("setting alert state", "uid", alertRule.UID)
 	switch result.State {
 	case eval.Normal:
-		currentState.resultNormal(alertRule, result)
+		instance.resultNormal(alertRule, result)
 	case eval.Alerting:
-		currentState.resultAlerting(alertRule, result)
+		instance.resultAlerting(alertRule, result)
 	case eval.Error:
-		currentState.resultError(alertRule, result)
+		instance.resultError(alertRule, result)
 	case eval.NoData:
-		currentState.resultNoData(alertRule, result)
+		instance.resultNoData(alertRule, result)
 	case eval.Pending: // we do not emit results with this state
 	}
 
+	// Set reason iff: state is not Pending, reason is different than state, reason is not Alerting or Normal
+	instance.StateReason = ""
+	if instance.State != eval.Pending &&
+		instance.State != result.State &&
+		result.State != eval.Normal &&
+		result.State != eval.Alerting {
+		instance.StateReason = result.State.String()
+	}
```

Comment:
I wonder if we can make it more explicit, i.e.  `(result.State == eval.Error || result.State == eval.NoData) && instance.State != result.State`